webapp/scripts/modules/query.js

```diff
@@ -0,0 +1,183 @@
+const fs = require('fs')
+const { addVector, dbSelectAll, getAllColumns, getLayers, getUnivarBounds, gpkgOut, grass, initMapset, listUserVector, mapsetExists, remove } = require('../grass')
+const { checkWritableDir, filterDefaultLayers, getFilesOfType, mergePDFs, psToPDF, textToPS } = require('../helpers')
+const { query: messages } = require('../messages.json')
+
+const GEOSERVER = `${process.env.GEOSERVER_DATA_DIR}/data`
+const GRASS = process.env.GRASS_DIR
+const OUTPUT = process.env.OUTPUT_DIR
+
+const QUERY_RESOLUTION = 0.00002
+const QUERY_MAP_NAME = 'query_map'
+const QUERY_RESULT_NAME = 'query_result'
+
+module.exports = class {
+  constructor() {
+    this.mapset = 'query'
+  }
+
+  launch() {
+    checkWritableDir(GEOSERVER)
+    checkWritableDir(OUTPUT)
+
+    if (!mapsetExists('PERMANENT')) {
+      return messages[7]
+    }
+
+    initMapset(this.mapset)
+
+    // set selection as the query area
+    grass(this.mapset, `g.copy vector=selection@PERMANENT,selection --overwrite`)
+    this.queryArea = 'selection'
+
+    const allVector = listUserVector()
+    const allGpkg = getFilesOfType('gpkg', GEOSERVER).filter(filterDefaultLayers)
+
+    /**
+     * check and add all layers from layer switcher
+     * FIXME: mapFile name and layer name differs, which makes it hard to check if file has already been imported
+     * ideal solution: 1. make sure all map files contains only one layer; 2. make sure all mapFile names align with its layer's name
```

Comment:
- If that's because you fixed special characters in the filename, use the fixed name for checking
- Multi-layered datasets are tricky... I think we have to either avoid them completely (i.e., import only a one layer) or construct a composite layer name like `${fileName}~${layerName}`

Reply:
The mapFile name and layer name differs because the person who named the files didn't make sure it aligns with the layer name. We can change the file names and the workspace file locations in another branch.

Reply:
1. We don't care about the file names or layer names–we only need to make sure we have a distinct mapping from one into the other.
2. OK, let's ignore multi-layered files for now. Please disable the corresponding code parts.